File: Assets/common/script/ship/TargettableObject.js
#pragma strict

class TargettableObject  extends MonoBehaviour{
	var objectName : String;
	var stateText : String;
	var targetted : boolean = false;
	var targetId : int = 0;
	var scanCode : int = 1234;	//id to scan to target this missile
	
	
	var health : float = 1.0f;
	
	
	var statNames : String[];
	var statValues : float[];
	
	
	
	var baseDamage : float = 5.0f;
	var exploding : boolean  = false;
	var trackingPlayer : boolean = false;	//is this tracking the players movements?
	var damageable : boolean = true;
	var targettable : boolean  = true;
	var grappleable : boolean = false;
		
	
	var colour : Color;
	var visibleAtClient : boolean = true;	//is the object visible at the client end?
	var visibleAtTactical : boolean = false;	//visible for tactical?
	var visibleAtPilot : boolean = true;		//visible for pilot?
	var highlighted : boolean = false;
	
	
	
	function Start () {
		targetId = gameObject.GetHashCode();
		if(statNames == null || statValues == null){
			statNames = new String[2];
			statValues = new float[2];
		}
		statNames[0] = "health";
		
		statValues[0] = 1;
		
	}
	
	/* object statistic handlers, "stats" are things sent to the radar/tactical comp when in range*/	
	function getStatIdFromName(n : String) : int{
		for(var i = 0; i < statNames.Length; i++){
			if(statNames[i] == n){
				return i;
			}
		} 
		return -1;
	}
	
	function getStatFromName(s : String) : float {
		var ind = getStatIdFromName(s);
		if(ind != -1){
			return statValues[ind];
		}
		return -10000.0f;
	} 
	
	function setStatFromName(s : String, val : float){
		var ind = getStatIdFromName(s);
		if(ind != -1){
			statValues[ind] = val;
		}
	}
		
	
	function Update () {
	
	}
	
	function GetShot(damage : float){}
	
	
	function explode() : IEnumerator{}
	
	function onTarget(){}
	function onUnTarget(){}
	
	function gotGrappled() {}
	function releaseGrapple(){}
}

File: Assets/common/script/ship/TargettableObject.cs
using UnityEngine;
using System;
using System.Collections;

[System.Serializable]
public class TargettableObject: MonoBehaviour{
	public string objectName;
	public string stateText;
	public bool targetted = false;
	public int targetId = 0;
	public int scanCode = 1234;	//id to scan to target this missile
	
	
	public float health = 1.0f;
	
	
	public string[] statNames;
	public float[] statValues;
	
	
	
	public float baseDamage = 5.0f;
	public bool exploding  = false;
	public bool trackingPlayer = false;	//is this tracking the players movements?
	public bool damageable = true;
	public bool targettable  = true;
	public bool grappleable = false;
		
	
	public Color colour;
	public bool visibleAtClient = true;	//is the object visible at the client end?
	public bool visibleAtTactical = false;	//visible for tactical?
	public bool visibleAtPilot = true;		//visible for pilot?
	public bool highlighted = false;
	
	public bool doNotInterpolate = false;	//do not interpolate the position of this object at the radar client end
	
	
	
	public virtual void Start() {
		targetId = gameObject.GetHashCode();
		if(statNames == null || statValues == null){
			statNames = new String[2];
			statValues = new float[2];
		}
		statNames[0] = "health";
		
		statValues[0] = 1.0f;
		
	}
	
	public void setPosition(Vector3 newPosition){
		transform.position = newPosition;
		doNotInterpolate = true;	//this will be cleared after the next radar update. It prevents targets jumping around the radar
									//and looking crappy
	}
	
	/* object statistic handlers, "stats" are things sent to the radar/tactical comp when in range*/	
	public int getStatIdFromName(string n){
		for(int i = 0; i < statNames.Length; i++){
			if(statNames[i] == n){
				return i;
			}
		} 
		return -1;
	}
	
	public float getStatFromName(string s) {
		int ind = getStatIdFromName(s);
		if(ind != -1){
			return statValues[ind];
		}
		return -10000.0f;
	} 
	
	public void setStatFromName(string s,float val){
		int ind = getStatIdFromName(s);
		if(ind != -1){
			statValues[ind] = val;
		}
	}
		
	
	public virtual void Update() {
	
	}
	
	public virtual void GetShot(float damage){}
	
	
	public virtual IEnumerator explode(){
		yield return null;
	}
	
	public virtual void onTarget(){}
	public virtual void onUnTarget(){}
	
	public void gotGrappled() {}
	public void releaseGrapple(){}
}